### ISSUE TITLE
native encryption + cache device = (ARC?) corruption

### ISSUE THREAD

**Issue by mklemm2** [zfsonlinux/zfs#6742]
### System information
Type                                | Version/Name
  ---                                  |     --- 
Distribution Name       | Arch Linux
Distribution Version    | 
Linux Kernel                 | 4.12.8
Architecture                 | x86-64
ZFS Version                  | 0.7.0-1 from https://github.com/tcaputi/zfs/commit/5aef9bedc801830264428c64cd2242d1b786fd49
SPL Version                  | 0.7.0-1 from https://github.com/tcaputi/zfs/commit/5aef9bedc801830264428c64cd2242d1b786fd49

### Describe the problem you're observing

On an encrypted raidz dataset with cache device, the content of files appears corrupted after a short time. After exporting and importing the pool, the content is okay, so the data is not really corrupted on-disk.

Does not happen with unencrypted datasets or encrypted datasets in a pool without cache device.

### Describe how to reproduce the problem

```
$ vagrant ssh arch
Last login: Mon Oct  9 13:17:43 2017 from 10.0.2.2
[vagrant@arch ~]$ sudo zpool import storage
[vagrant@arch ~]$ sudo zfs load-key storage
Enter passphrase for 'storage':
[vagrant@arch ~]$ sudo zfs mount -a
[vagrant@arch ~]$ cd /storage/
[vagrant@arch storage]$ sudo chown vagrant .
[vagrant@arch storage]$ dd if=/dev/urandom of=test1.dat bs=1M count=512
512+0 records in
512+0 records out
536870912 bytes (537 MB, 512 MiB) copied, 4.23188 s, 127 MB/s
[vagrant@arch storage]$ cp test1.dat test2.dat
[vagrant@arch storage]$ cmp test1.dat test2.dat
[vagrant@arch storage]$ cmp test1.dat test2.dat
[vagrant@arch storage]$ cmp test1.dat test2.dat
test1.dat test2.dat differ: byte 1, line 1
[vagrant@arch storage]$ zpool status
  pool: storage
 state: ONLINE
  scan: none requested
config:

	NAME        STATE     READ WRITE CKSUM
	storage     ONLINE       0     0     0
	  raidz1-0  ONLINE       0     0     0
	    sdb     ONLINE       0     0     0
	    sdc     ONLINE       0     0     0
	    sdd     ONLINE       0     0     0
	logs
	  sdf       ONLINE       0     0     0
	cache
	  sde       ONLINE       0     0     0

errors: No known data errors
[vagrant@arch storage]$ cd
[vagrant@arch ~]$ sudo zpool export storage
[vagrant@arch ~]$ sudo zpool import storage
[vagrant@arch ~]$ sudo zfs load-key storage
Enter passphrase for 'storage':
[vagrant@arch ~]$ sudo zfs mount -a
[vagrant@arch ~]$ cd -
/storage
[vagrant@arch storage]$ cmp test1.dat test2.dat
[vagrant@arch storage]$ cmp test1.dat test2.dat
[vagrant@arch storage]$ cmp test1.dat test2.dat
[vagrant@arch storage]$ cmp test1.dat test2.dat
[vagrant@arch storage]$ cmp test1.dat test2.dat
[vagrant@arch storage]$ cmp test1.dat test2.dat
[vagrant@arch storage]$ cmp test1.dat test2.dat
[vagrant@arch storage]$ cmp test1.dat test2.dat
[vagrant@arch storage]$ cmp test1.dat test2.dat
test1.dat test2.dat differ: byte 1, line 1
```

**Comment by mklemm2** [comment]
The issue also happened with https://github.com/zfsonlinux/zfs/commit/269db7a4b but seems to be fixed in https://github.com/zfsonlinux/zfs/commits/master

**Comment by behlendorf** [comment]
@tcaputi can you look in to this.

**Comment by tcaputi** [comment]
@mkjorling Would you mind trying this again with #6595 applied? I fixed a few L2ARC bugs there and this may already be resolved. I will try to reproduce it locally in the meantime

**Comment by tcaputi** [comment]
My testing indicates that this is fixed already in #6595. Reopen the issue if this isn't true.
